File: Samples/Java/Console/src/main/java/io/cratis/chronicle/samples/console/EmployeeDetails.java
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

package io.cratis.chronicle.samples.console;

import io.cratis.chronicle.projections.FromEvent;
import io.cratis.chronicle.projections.SetFrom;
import io.cratis.chronicle.readModels.ReadModel;

/**
 * Model-bound projection artifact discovered by the Kotlin client.
 *
 * Each {@link FromEvent} annotation on the class declares which event types contribute to this read model.
 * Properties are auto-mapped by name unless overridden with {@link SetFrom}.
 */
@ReadModel
@FromEvent(eventType = EmployeeHired.class)
@FromEvent(eventType = EmployeeAddressSet.class)
@FromEvent(eventType = EmployeePromoted.class)
@FromEvent(eventType = EmployeeMoved.class)
public class EmployeeDetails {
    private String id = "";
    private String firstName = "";
    private String lastName = "";
    @SetFrom(propertyPath = "newTitle", eventType = EmployeePromoted.class)
    private String title = "";
    private String address = "";
    private String city = "";
    private String zipCode = "";
    private String country = "";

    public EmployeeDetails() {}

    public EmployeeDetails(String id, String firstName, String lastName, String title,
                          String address, String city, String zipCode, String country) {
        this.id = id;
        this.firstName = firstName;
        this.lastName = lastName;
        this.title = title;
        this.address = address;
        this.city = city;
        this.zipCode = zipCode;
        this.country = country;
    }

    public String getId() { return id; }
    public void setId(String id) { this.id = id; }

    public String getFirstName() { return firstName; }
    public void setFirstName(String firstName) { this.firstName = firstName; }

    public String getLastName() { return lastName; }
    public void setLastName(String lastName) { this.lastName = lastName; }

    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }

    public String getAddress() { return address; }
    public void setAddress(String address) { this.address = address; }

    public String getCity() { return city; }
    public void setCity(String city) { this.city = city; }

    public String getZipCode() { return zipCode; }
    public void setZipCode(String zipCode) { this.zipCode = zipCode; }

    public String getCountry() { return country; }
    public void setCountry(String country) { this.country = country; }
}

File: Samples/Kotlin/Console/src/main/kotlin/io/cratis/chronicle/samples/console/EmployeeDetails.kt
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

package io.cratis.chronicle.samples.console

import io.cratis.chronicle.projections.FromEvent
import io.cratis.chronicle.projections.FromEventSourceId
import io.cratis.chronicle.projections.Increment
import io.cratis.chronicle.projections.SetFrom
import io.cratis.chronicle.readModels.ReadModel

/**
 * Model-bound projection artifact discovered by the Kotlin client.
 *
 * Each [FromEvent] annotation on the class declares which event types contribute to this read model.
 * Properties are auto-mapped by name unless overridden with [SetFrom].
 */
@ReadModel
@FromEvent(EmployeeHired::class)
@FromEvent(EmployeeAddressSet::class)
@FromEvent(EmployeePromoted::class)
@FromEvent(EmployeeMoved::class)
data class EmployeeDetails(
    /**
     * The read model's key, taken from the event source id of every event it observes.
     *
     * The sink stores each instance under its `id`, so a projection whose read model has one has to
     * say where it comes from - leave it empty and every employee lands on the same document.
     *
     * The kernel does not act on this yet (https://github.com/Cratis/Chronicle/issues/3924), which
     * is why pressing 'L' lists one merged employee rather than three. The reducer-backed
     * EmployeeState next door stamps its own key and does show all three.
     */
    @FromEventSourceId val id: String = "",
    val firstName: String = "",
    val lastName: String = "",
    @SetFrom("newTitle", EmployeePromoted::class) val title: String = "",
    val address: String = "",
    val city: String = "",
    val zipCode: String = "",
    val country: String = "",
    /** Bumped by one every time this employee is promoted — demonstrates [Increment]. */
    @Increment(EmployeePromoted::class) val promotionCount: Int = 0
)
